src/bundle_info.c

```diff
@@ -0,0 +1,570 @@
+					print(" - %s (optional)\n", included_bundle);
+				}
+				iter = iter->next;
+			}
+		}
+	}
+
+	/* print indirect includes (if any) */
+	iter = list_head(bundle_info->indirect_includes);
+	if (iter) {
+		print("\nIndirectly included bundles (%d):\n", list_len(bundle_info->indirect_includes));
+		while (iter) {
+			included_bundle = iter->data;
+			print(" - %s\n", included_bundle);
+			iter = iter->next;
+		}
+	}
+}
+
+static void print_files(struct bundle_info *bundle_info)
+{
+	struct manifest *bundle_manifest = bundle_info->manifest;
+	struct list *iter;
+	struct file *file;
+	long count = 0;
+
+	print("\nFiles in bundle");
+
+	if (cmdline_option_includes) {
+		iter = list_head(bundle_info->files);
+		print(" (includes dependencies):\n");
+	} else {
+		iter = list_head(bundle_manifest->files);
+		print(":\n");
+	}
+
+	while (iter) {
+		count++;
+		file = iter->data;
+		print(" - %s\n", file->filename);
+		iter = iter->next;
+	}
+
+	print("\nTotal files: %d\n", count);
+}
+
+static enum swupd_code get_indirect_includes(struct manifest *mom, struct list *recurse_includes, struct list *direct_includes, struct list **indirect_includes, bool from_recursion)
+{
+	struct list *iter;
+	struct file *file = NULL;
+	struct manifest *manifest = NULL;
+	char *included_bundle;
+
+	iter = list_head(recurse_includes);
+	while (iter) {
+
+		included_bundle = iter->data;
+		iter = iter->next;
+
+		/* if the bundle has already been added to the direct includes list do not
+		 * add it as indirect include */
+		if (from_recursion) {
+			if (list_search(direct_includes, included_bundle, list_strcmp)) {
+				continue;
+			}
+		}
+
+		file = search_bundle_in_manifest(mom, included_bundle);
+		if (!file) {
+			error("Bundle %s could not be found in the MoM\n", included_bundle);
+			return SWUPD_INVALID_BUNDLE;
+		}
+
+		manifest = load_manifest(file->last_change, file, mom, true, NULL);
+		if (!manifest) {
+			error("Manifest for bundle %s could not be loaded\n", included_bundle);
+			return SWUPD_COULDNT_LOAD_MANIFEST;
+		}
+
+		/* only add to the list those includes gotten by recursion */
+		if (from_recursion) {
+			*indirect_includes = list_prepend_data(*indirect_includes, strdup_or_die(included_bundle));
+		}
+
+		if (manifest->includes) {
+			get_indirect_includes(mom, manifest->includes, direct_includes, indirect_includes, true);
+		}
+
+		if (manifest->optional) {
+			get_indirect_includes(mom, manifest->optional, direct_includes, indirect_includes, true);
+		}
+
+		free_manifest(manifest);
+	}
+
+	return SWUPD_OK;
+}
+
+static enum swupd_code get_included_bundles(struct bundle_info *bundle_info, struct manifest *mom)
+{
+	struct manifest *bundle_manifest = bundle_info->manifest;
+	struct list *direct_includes = NULL;
+	struct list *indirect_includes = NULL;
+	enum swupd_code ret;
+
+	/* get a list of all includes (direct and indirect includes) */
+	direct_includes = list_clone_and_concat(bundle_manifest->includes, bundle_manifest->optional);
```

Comment:
because if I just concatenate manifest->includes with manifest->optional then both those lists are going to become one, and I need to be able to read them separately at a later step.

Reply:
Actually I am overcomplicating things, I think I found a simpler way to get the info. I'll push an update soon.